File: app/javascripts/vpago/vpago_payments/user_informers/firebase.js
import { initializeApp } from "firebase/app";
import { getFirestore, doc, onSnapshot, setDoc } from "firebase/firestore";
import QueueProcessor from "../../../queue_processor.js";

async function listenToProcessingState({
  firebaseConfigs,
  documentReferencePath,
  onPaymentIsProcessing,
  onOrderIsProcessing,
  onOrderIsCompleted,
  onOrderProcessFailed,
  onPaymentIsRefunded,
  onPaymentProcessFailed,
  onCompleted,
}) {
  const app = initializeApp(firebaseConfigs);
  const db = getFirestore(app);

  const documentRef = doc(db, documentReferencePath);
  await setDoc(documentRef, { listening: true }, { merge: true });

  const queueProcessor = new QueueProcessor();

  onSnapshot(documentRef, (doc) => {
    let documentData = doc.data();

    let messageCode = documentData["message_code"];
    let orderState = documentData["order_state"];
    let paymentState = documentData["payment_state"];
    let processing = documentData["processing"] === true;
    let reasonCode = documentData["reason_code"];
    let reasonMessage = documentData["reason_message"];

    let orderCompleted = orderState === "complete";
    if (orderCompleted) {
      queueProcessor.queueStateChange({
        minDelayInMs: 1500,
        callback: async () => {
          await onCompleted(
            orderState,
            paymentState,
            reasonCode,
            reasonMessage
          );
        },
      });
      return;
    }

    switch (messageCode) {
      case "payment_is_processing":
        queueProcessor.queueStateChange({
          minDelayInMs: 1500,
          callback: async () => {
            await onPaymentIsProcessing(
              orderState,
              paymentState,
              processing,
              reasonCode,
              reasonMessage
            );
          },
        });
        break;
      case "order_is_processing":
        queueProcessor.queueStateChange({
          minDelayInMs: 1500,
          callback: async () => {
            await onOrderIsProcessing(
              orderState,
              paymentState,
              processing,
              reasonCode,
              reasonMessage
            );
          },
        });
        break;
      case "order_is_completed":
        queueProcessor.queueStateChange({
          minDelayInMs: 1500,
          callback: async () => {
            await onOrderIsCompleted(
              orderState,
              paymentState,
              processing,
              reasonCode,
              reasonMessage
            );
          },
        });
        break;
      case "order_process_failed":
        queueProcessor.queueStateChange({
          minDelayInMs: 1500,
          callback: async () => {
            await onOrderProcessFailed(
              orderState,
              paymentState,
              processing,
              reasonCode,
              reasonMessage
            );
          },
        });
        break;
      case "payment_is_refunded":
        queueProcessor.queueStateChange({
          minDelayInMs: 1500,
          callback: async () => {
            await onPaymentIsRefunded(
              orderState,
              paymentState,
              processing,
              reasonCode,
              reasonMessage
            );
          },
        });
        break;
      case "payment_process_failed":
        queueProcessor.queueStateChange({
          minDelayInMs: 1500,
          callback: async () => {
            await onPaymentProcessFailed(
              orderState,
              paymentState,
              processing,
              reasonCode,
              reasonMessage
            );
          },
        });
        break;
      default:
        break;
    }
  });
}

window.listenToProcessingState = listenToProcessingState;

File: app/services/vpago/user_informers/firebase.rb
require 'google/cloud/firestore'

module Vpago
  module UserInformers
    class Firebase
      attr_accessor :order

      def initialize(order)
        @order = order
      end

      def payment_is_processing(processing:) = notify(:payment_is_processing, processing)
      def payment_is_retrying(processing:) = notify(:payment_is_retrying, processing)
      def order_is_processing(processing:) = notify(:order_is_processing, processing)
      def order_is_completed(processing:) = notify(:order_is_completed, processing)
      def order_process_failed(processing:, reason_code:, reason_message: nil) = notify(:order_process_failed, processing, reason_code, reason_message)
      def payment_process_failed(processing:, reason_code:, reason_message: nil) = notify(:payment_process_failed, processing, reason_code, reason_message)

      def notify(message_code, processing, reason_code = nil, reason_message = nil)
        order.reload

        data = {
          processing: processing,
          message_code: message_code,
          reason_code: reason_code,
          reason_message: reason_message,
          order_state: order.state,
          payment_state: order.payment_state,
          updated_at: Time.current
        }.compact

        firestore_reference.set(data, merge: true)
        firestore_reference.col('histories').doc(message_code).set(data)
      end

      def firestore_reference
        order_created_date = order.created_at.strftime('%Y-%m-%d')
        @firestore_reference ||= firestore.col('statuses')
                                          .doc('cart')
                                          .col(order_created_date)
                                          .doc(order.number)
      end

      def document_reference_path
        firestore_reference.path.split('/documents').last
      end

      def firestore
        @firestore ||= Google::Cloud::Firestore.new(project_id: service_account[:project_id], credentials: service_account)
      end

      def service_account
        @service_account ||= Rails.application.credentials.cloud_firestore_service_account
      end
    end
  end
end
